[1단계 - 자동차 경주 구현] 알린(장원영) 미션 제출합니다.

OzRagwort opened the pull request:
안녕하세요 반갑습니다 에단! 저는 알린입니다. 

저는 티키와 함께 페어 프로그래밍을 했는데 다른 사람과 같은 코드를 보는 게 신기하고 재미있었습니다. 

서로 잘 알지 못하던 것들을 공유하고 생각이 다를 때 토론하면서 더 나은 코드를 작성할 수 있는 것 같아서 좋았습니다.

리뷰가 처음이라 설레고 떨리네요.. 잘 부탁드립니다!!!

wyRagwort committed:
docs(README): 문자열 덧셈 계산기 구현 기능 목록 추가

wyRagwort committed:
test(calculator): null 값 또는 빈 문자열을 입력할 경우 0을 반환 테스트

wyRagwort committed:
feat(calculator): null 값 또는 빈 문자열을 입력할 경우 0을 반환

wyRagwort committed:
test(calculator): 입력 문자열을 구분자로 나누는 기능

wyRagwort committed:
feat(calculator): 입력 문자열을 구분자로 나누는 기능

wyRagwort committed:
test(calculator): 문자열 배열을 숫자 배열로 바꾸는 기능 테스트

wyRagwort committed:
feat(calculator): 문자열 배열을 숫자 배열로 바꾸는 기능 추가

wyRagwort committed:
test(calculator): 숫자를 모두 더하는 기능 테스트

wyRagwort committed:
test(matcher): Matcher의 group에 대한 학습 테스트

wyRagwort committed:
test(calculator): 원하는 구분자를 설정하는 기능 테스트

wyRagwort committed:
feat(calculator): 원하는 구분자를 설정하는 기능 추가

wyRagwort committed:
test(calculator): 입력이 숫자가 아닌 경우 예외 처리 테스트

wyRagwort committed:
feat(calculator): 입력이 숫자가 아닌 경우 예외 처리 기능 추가

wyRagwort committed:
test(calculator): 입력이 음수인 경우 예외 처리 테스트

wyRagwort committed:
feat(calculator): 입력이 음수인 경우 예외 처리 기능 추가

wyRagwort committed:
test(calculator): 입력이 하나인 경우 그대로 반환하는 기능 테스트

wyRagwort committed:
refactor(calculator): 매직 넘버 값을 상수로 리팩터링

wyRagwort committed:
refactor(matcher): matcherTest를 studyTest 패키지로 이동

wyRagwort committed:
refactor(calculator): 입력이 null 또는 빈 문자열인 것을 검증하는 메소드 분리

wyRagwort committed:
test(calculator): 숫자를 모두 더하는 기능 추가

wyRagwort committed:
refactor(calculator): 메소드명 toInteger를 toIntegers로 수정

wyRagwort committed:
Merge branch 'calcaulator' into ozragwort

yh20studio committed:
docs(README): 자동차 경주 게임 구현 기능 목록 추가

yh20studio committed:
test(racingCar): 자동차 이름을 입력받는 기능 테스트

yh20studio committed:
feat(racingCar): 자동차 이름을 입력받는 기능 추가

yh20studio committed:
test(racingCar): 자동차 이름을 나누는 기능 테스트

yh20studio committed:
feat(racingCar): 자동차 이름을 나누는 기능 추가

yh20studio committed:
refactor(racingCar): 불필요한 static 제어자 제거

yh20studio committed:
test(application): 정상 동작을 확인하는 통합 테스트 추가

yh20studio committed:
refactor(racingCar): 메소드 명 readCarNames를 readNextLine로 수정

yh20studio committed:
feat(application):  main 함수 추가

yh20studio committed:
test(attempt): 시도할 회수를 설정하는 기능 테스트

yh20studio committed:
feat(attempt): 시도할 회수를 설정하는 기능 추가

yh20studio committed:
test(position): 자동차의 위치 Value Object 테스트

yh20studio committed:
feat(position): 자동차의 위치를 Value Object 추가

yh20studio committed:
test(carName): 자동차의 이름 Value Object 테스트

yh20studio committed:
feat(carName): 자동차의 이름 Value Object 추가

yh20studio committed:
test(car) : 자동차 Value Object 테스트

yh20studio committed:
feat(car): 자동차 Value Object 추가

yh20studio committed:
test(cars): 자동차들의 일급 컬렉션 테스트

yh20studio committed:
feat(cars): 자동차들의 일급 컬렉션 추가

yh20studio committed:
refactor(racingCar): cars 일급 컬렉션 리팩토링

yh20studio committed:
test(car): 자동차가 전진하는 기능 테스트

yh20studio committed:
feat(car): 자동차가 전진하는 기능 추가

yh20studio committed:
test(car): 자동차의 실행 결과를 반환하는 기능 테스트

yh20studio committed:
feat(car): 자동차의 실행 결과를 반환하는 기능 추가

yh20studio committed:
test(cars): 시도할 회수만큼 경주하는 기능 테스트

yh20studio committed:
feat(cars): 시도할 회수만큼 경주하는 기능 추가

yh20studio committed:
refactor(car): move 메서드 이동 결과 값 반환

yh20studio committed:
test(cars): 우승자를 반환하는 기능 테스트

yh20studio committed:
feat(cars): 우승자를 반환하는 기능 추가

yh20studio committed:
refactor(winners): 우승자들의 일급 컬렉션 추가

wyRagwort committed:
feat(view): UI 출력 기능 추가

wyRagwort committed:
feat(inputView): 자동차 이름과 시도 회수를 입력하지 않은 경우 예외 발생 기능

wyRagwort committed:
feat(racingCar): 자동차 이름 중 공백이 있는 경우 예외 발생 기능

wyRagwort committed:
feat(carName): 자동차 이름이 5자리를 초과하는 경우 예외 발생 기능

wyRagwort committed:
feat(cars): 자동차 이름이 중복 되었을 때 예외 발생 기능

wyRagwort committed:
feat(attempt): 시도 회수가 숫자가 아닌 경우 예외 발생 기능

wyRagwort committed:
feat(attempt): 시도 회수가 음수인 경우 예외 발생 기능

wyRagwort committed:
refactor: 매직 넘버를 상수로 변경

wyRagwort committed:
Merge branch 'exception' into ozragwort

wyRagwort committed:
Merge branch 'refactor' into ozragwort

# Conflicts:
#	src/main/java/racingcar/RacingCar.java
#	src/main/java/racingcar/view/InputView.java
#	src/main/java/racingcar/vo/Attempt.java
#	src/main/java/racingcar/vo/CarName.java
#	src/main/java/racingcar/vo/Cars.java

wyRagwort committed:
refactor: 메소드 및 변수명이 의도를 제대로 나타낼 수 있도록 수정

wyRagwort committed:
refactor: 테스트 코드에서 DisplayName 추가

wyRagwort committed:
refactor: 메소드 순서를 일관성있게 변경

wyRagwort committed:
fix(carName): equals와 hashCode Override

wyRagwort committed:
Merge branch 'refactor' into ozragwort

Laterality commented on src/main/java/racingcar/util/RandomUtil.java:
```diff
@@ -0,0 +1,26 @@
+
+  private static void validRange(int randomNumber) {
+    if (randomNumber > MAXIMUM || randomNumber < MINIMUM) {
+      throw new IllegalArgumentException(RANGE_OVER_ERROR_MESSAGE);
```
`IllegalArgumentException`은 클라이언트가 전달한 인자가 올바르지 않은 경우에 던지는 것이 좋은데요.

사용자에게 노출된 `generateRandomNumber`는 아무런 인자도 없다보니 사용자 입장에서는 아무런 인자 없이 호출한 메서드가 `IllegalArgumentException`을 던진다면 당황스러울 수 있어보입니다.

로직을 수행하다 예기치 못한 상태가 되는 경우에 던질만한 다른 예외가 있을까요?

OzRagwort replied:
예외 처리를 하면서 어떤 문제에서 Exception을 발생시켜야 하는지 깊게 고민해보지 않았던것 같습니다. 

어떤 값이 잘못되면 아무런 의심없이 `IllegalArgumentException`를 썼던것 같아요. 

이 부분은 랜덤하게 생성한 값이 원하는 범위에 있지 않을 때 Exception이 발생하게 되는데 "예기치 못한 상태"가 되는 경우라는 것에서 힌트를 얻어 `IllegalStateException`를 써봤습니다. 

처음으로 RuntimeException에 어떤 Exception들이 있는지 봤는데 참 많은 종류가 있는것을 알게되었습니다. 

앞으로 예외를 던질 때 이 상황에서 이 예외가 맞는지 한번 더 생각해보게 될 것 같아요!

Laterality commented on src/main/java/racingcar/vo/Attempt.java:
```diff
@@ -0,0 +1,52 @@
+    attempt--;
+  }
+
+  private int valid(String attempt) {
```
일반적으로 널리 사용되는 코드 작성 규칙 중 하나인 [Google Style Guide](https://google.github.io/styleguide/javaguide.html#s5.2.3-method-names)에서는 메서드 이름이 동사로 시작해야 한다고 밝히고 있는데요. `validate`나 `check`같은 이름은 어떨까요?

OzRagwort replied:
google java 컨벤션을 다시한번 읽다가 궁금한 점이 생겨서 질문드립니다. 

네이버 java 컨벤션에서는 메소드 이름이 동사/전치사로 시작한다고 명시가 되어있는데 google java 컨벤션에서는 동사/동사구로 시작한다고만 되어있는것 같습니다. 

보통은 동사로 시작하지만 toString()이나 of()와 같이 전치사로 시작되는 메소드들도 많은데 이렇게 전치사로 시작하는 메소드를 실제로 만들어 사용하는 경우도 있는지 궁금합니다!

Laterality replied:
네, 실제로도 동사 외의 단어를 사용하기도 합니다. `Optional<T>` 클래스에도 `ifPresent(Consumer<? super T> action)`와 같은 메서드가 있지요. 함수를 인자로 받기 때문에 앞에 `do`가 생략된 거라고도 볼 수 있겠네요(`doIfPresent`). 동사 외의 단어도 사용할 수 있지만, 값 형태의 인자를 받아 반환하는 흔히 볼 수 있는 형식의 메서드는 대부분 동사로도 메서드를 표현하는 데 큰 무리가 없다보니 가급적 동사 사용을 지향하시면 좋겠습니다.

Laterality commented on src/main/java/racingcar/vo/Car.java:
```diff
@@ -0,0 +1,48 @@
+  }
+
+  @Override
+  public int compareTo(Car car) {
```
👍

OzRagwort replied:
감사합니다! 👍

Laterality commented on src/main/java/racingcar/vo/Cars.java:
```diff
@@ -0,0 +1,66 @@
+    cars.add(car);
+  }
+
+  public String repeatRaceBy(Attempt attempt) {
```
반환 타입이 `String`인 이유는 뷰에서 그대로 출력하기 위함일까요?

만약 출력 형식이 바뀌어야 한다면 어느 `OutputView`와 `Cars` 중 어디가 바뀌어야 할까요?

OzRagwort replied:
네 String으로 반환한 이유는 그대로 출력하기 위해서였습니다. 그런데 출력 형식이 바뀌어야 한다면 OutputView와 Cars중 어디가 바뀌어야 할까를 생각해보니 OutputView가 바뀌어야 되는 게 맞다고 느꼈습니다. 

Position에서도 출력 형식에 맞추어 toString()을 했는데 이렇게 되면 VO가 View의 세부 사항을 알아야 하고 수정할 때도 View가 아닌 VO를 수정해야 하기 때문에 문제가 될 것 같다고 느꼈습니다. 

그래서 입출력에 대한 것은 최대한 View에서 처리하고, VO는 해당 객체의 로직만을 담당하도록 수정하려고 노력해봤습니다. 

저는 RoundResults를 만들어 이곳에 매 시도 횟수 마다의 자동차 이름과 이동 거리를 저장하고 반환하도록 수정했습니다. 

OutputView에서는 이 RoundResults에서 시도 회수마다 자동차 이름과 이동 거리를 얻어 정해진 출력 형식으로 출력하도록 수정하였습니다. 

지금의 구현 방식에서 고쳐야 할 점이 있다면 힌트 부탁드립니다!

Laterality replied:
좋은 방향입니다 👍

Laterality commented on src/main/java/racingcar/vo/Cars.java:
```diff
@@ -0,0 +1,66 @@
+    }
+  }
+
+  public boolean isSize(int size) {
```
1. 일반적으로 "isXXX" 메서드는 "객체가 XXX인지 확인한다"는 의미를 나타냅니다. 사이즈가 같은지 확인하는 역할이라면 `matchSize` 같은 이름은 어떨까요?
2. `isSize`는 테스트에서만 사용되는 메서드로 보이는데, 이 메서드 없이 테스트할 수 있는 방법은 없을까요?

OzRagwort replied:
`isSize`가 없다면 테스트가 절대 불가능한지 고민해보았는데 `toString`과 `assertThat().contains()`를 이용하여 테스트를 할 수는 있을 것 같습니다. 

하지만 `toString`이 변경되면 테스트가 실패하게 될 것 같아서 제대로 테스트를 한 것인지는 확신이 없습니다. 

또 `extracting`를 이용하여 테스트해봤는데 실패했습니다. `extracting`를 이용한 테스트는 조금 더 공부해서 다시 시도해보려고 합니다.

Laterality replied:
이펙티브 자바의 Item 12를 보시면 비슷한 내용을 다루고 있는데요(책을 보셔도 좋고, 검색해보시면 정리된 내용이 많이 있어요!). 
만약 상태를 노출하지 않음으로 인해 테스트가 더 어려워진다면 노출하는 것도 답이 될 수 있겠죠?

Laterality commented on src/main/java/racingcar/vo/Position.java:
```diff
@@ -0,0 +1,35 @@
+public class Position implements Comparable<Position> {
+
+  private static final int INIT_POSITION = 0;
+  private static final String POSITION_BAR = "-";
```
만일 출력 형식이 "-"에서 "+"로 바뀌어야 한다면 `OutputView` 대신(혹은 이와 함께) 이 클래스에도 변경이 필요할 텐데요. 모델 영역의 VO가 뷰의 세부사항을 알고 있는 것이 바람직할까요?

Laterality commented on src/main/java/racingcar/vo/Car.java:
```diff
@@ -0,0 +1,48 @@
+  }
+
+  public boolean move() {
+    if (isMoveForward(generateRandomNumber())) {
```
`Car`에서 직접 `RandomUtil`에서 숫자를 생성하고 있는데요. 

테스트할 때마다 다른 값이 나오다보니 `Car` 클래스에서 핵심적인 기능인데도 불구하고 테스트가 없네요. 

어떻게 하면 이 메서드를 테스트할 수 있을까요?

OzRagwort replied:
랜덤값이 들어간 `move` 메서드의 단위 테스트가 페어와 함께 가장 오래 고민했던 부분이었습니다. 

이 문제를 해결하기 위해 전략 패턴을 활용하여 인터페이스를 분리하여 테스트하기 좋은 메서드를 만드는 방법에 대한 글을 읽고 적용해보았습니다. 

https://tecoble.techcourse.co.kr/post/2020-05-17-appropriate_method_for_test_by_interface/

확실히 테스트가 더 편해진것이 느껴졌지만 아직은 전략 패턴이라는 개념이 완벽하게 이해가 되지는 않아 더 공부해야겠습니다!

Laterality replied:
좋은 방향입니다 👍

Laterality reviewed (changes requested):
안녕하세요 알린, 리뷰어 에단입니다.

VO로 각 요소의 제약 사항들을 분리하신 부분이 인상적이네요 👍 

추가로 보완하면 좋을 만한 부분들을 몇가지 짚어봤습니다. 참고 부탁드릴게요.

궁금하신 점이 있으시면 언제든 질문주세요!

wyRagwort committed:
refactor: 수정할 목록 TODO, FIXME로 정리

wyRagwort committed:
refactor: 메소드가 동사로 시작하도록 수정

- google java style 5.2.3 메소드 네이밍 규칙
https://google.github.io/styleguide/javaguide.html#s5.2.3-method-names

wyRagwort committed:
refactor: 메소드에서 잘못된 값을 가지게 될 경우 적절한 예외 처리를 하도록 수정

- IllegalStateException를 사용하여 예외 처리하도록 수정

wyRagwort committed:
refactor: 잘못된 숫자가 입력된 경우 적절한 예외 처리를 하도록 수정

wyRagwort committed:
refactor: Cars를 생성하는 책임을 Cars가 가지도록 수정

wyRagwort committed:
refactor: 모든 입,출력과 관련된 책임은 View가 가지도록 수정

wyRagwort committed:
refactor: Winners가 Car를 가지도록 수정

wyRagwort committed:
refactor: Car의 toString()에서 ","를 ":"으로 수정

wyRagwort committed:
refactor: 테스트에서만 사용되는 불필요한 메서드를 삭제

- 테스트에서만 사용되는 메서드를 지우고 toString()으로 테스트하도록 수정

wyRagwort committed:
test: 랜덤값을 이용하는 메소드에 대한 테스트를 추가

wyRagwort committed:
refactor: 테스트의 메소드 명을 모두 영어로 수정

wyRagwort committed:
Merge branch 'refactor' into ozragwort

wyRagwort committed:
fix(test): move() 테스트가 제대로 동작하도록 수정

OzRagwort commented:
안녕하세요 에단님

자세한 피드백 주셔서 감사합니다. 

주신 피드백을 이해해보려고 노력하면서 수정해봤습니다.

Laterality reviewed (approved):
안녕하세요 알린, 피드백을 꼼꼼하게 확인해주셨네요 👍 

이번 MR은 여기서 머지하도록 할게요. 다음 미션도 화이팅하세요!
